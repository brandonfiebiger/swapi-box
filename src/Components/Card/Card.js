import React from 'react';
import PropTypes from 'prop-types';

import './Card.css';

const Card = ({ facts, toggleFavorites, selected }) => {
  let fillColor = '#f2f2f2';
  const cardFacts = Object.keys(facts).map( (fact, index) => {
    selected ? fillColor = '#fac917' : fillColor = '#f2f2f2';
    if (fact === 'name') {
      return <li className={fact} key={index}>{facts[fact]}</li>
    } else if (fact === 'residents' && !facts[fact].length) {
      return <li key={index}><span>{fact}:</span> none</li>;
    } else {
      return <li className={fact} key={index}><span>{fact}:</span> {facts[fact]}</li>;
    }
  });

  return (
    <div 
      className='Card' 
      onClick={ () => toggleFavorites(facts) }>
      <svg 
        width='2rem'
        enableBackground="new 0 0 426.667 426.667" 
        viewBox="0 0 426.667 426.667" >
        <path 
          d="m213.333 10.441 65.916 133.576 147.418 
          21.419-106.667 103.974 25.173 
          146.816-131.84-69.316-131.848 69.316 
          25.182-146.816-106.667-103.974 147.409-21.419z" 
          fill={fillColor}/>
      </svg>
      <ul>
        {cardFacts}
      </ul>
    </div>
  );
};

export default Card;

Card.propTypes = {
  facts: PropTypes.objectOf(PropTypes.string),
  toggleFavorites: PropTypes.func,
  selected: PropTypes.bool
};