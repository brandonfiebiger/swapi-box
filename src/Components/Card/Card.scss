.Card {
  color: #fff;
  background: blanchedalmond;
  border: 5px solid cadetblue;
  border-radius: 10px;
  height: 20rem;
  min-width: 15rem;
  color: cadetblue;
  margin-left: 6rem;
  margin-bottom: 3rem;
  display: flex;
  flex-direction: column;
  opacity: 0.95;
  &:hover {
    opacity: 1;
    cursor: pointer;
  }
  &:active {
    opacity: 0.93;
  }
}

ul {
 height: 13rem;
//  line-height: 2.2rem;
}

.name {
  font-size: 1.5rem;
}

.residents {
  overflow: scroll;
  height: 3rem;
  line-height: 2rem;
  border: 1px solid cadetblue;
  margin-right: 2rem;
  height: 4rem;
  margin-bottom: 5rem;
}

@media screen and (max-width: 900px) {
  .Card {
    margin-left: 0;
  }
}