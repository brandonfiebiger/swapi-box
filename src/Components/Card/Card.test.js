import React from 'react';
import { shallow } from 'enzyme';

import Card from './Card';

describe('Card', () => {
  let facts;
  let mockToggleFavorites;
  let selected;

  beforeEach(() => {
    facts = {};
    mockToggleFavorites = jest.fn();
    selected = false;
  });

  it('should match the snapshot', () => {
    const wrapper = shallow(<Card facts={facts} 
      toggleFavorites={mockToggleFavorites} 
      selected={selected} />);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match the snapshot with a fact', () => {
    facts = {name: 'test'};
    const wrapper = shallow(<Card facts={facts} 
      toggleFavorites={mockToggleFavorites} 
      selected={selected} />);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match the snapshot with multiple facts', () => {
    facts = {name: 'test1', origin: 'the matrix'};
    const wrapper = shallow(<Card facts={facts} 
      toggleFavorites={mockToggleFavorites} 
      selected={selected} />);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call toggleFavorites with the correct parameters when clicked', () => {
    const wrapper = shallow(<Card facts={facts} 
      toggleFavorites={mockToggleFavorites} 
      selected={selected} />);
    
    wrapper.find('.Card').simulate('click');

    expect(mockToggleFavorites).toHaveBeenCalledWith(facts);
  });

  it('should change color of svg depending on if selected is true or false', () => {
    let wrapper = shallow(<Card facts={facts} 
      toggleFavorites={mockToggleFavorites} 
      selected={selected} />);
    
    expect(wrapper.find('path').prop('fill')).toEqual('#f2f2f2');

    selected = true;
    facts = {name: 'test1', origin: 'the matrix'};
    wrapper = shallow(<Card facts={facts} 
      toggleFavorites={mockToggleFavorites} 
      selected={selected} />);
      
    expect(wrapper.find('path').prop('fill')).toEqual('#fac917');
  });

});