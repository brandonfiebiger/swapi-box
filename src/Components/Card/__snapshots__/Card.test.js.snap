// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot 1`] = `"<div class=\\"Card\\"><svg width=\\"2rem\\" enable-background=\\"new 0 0 426.667 426.667\\" viewBox=\\"0 0 426.667 426.667\\"><path d=\\"m213.333 10.441 65.916 133.576 147.418  21.419-106.667 103.974 25.173  146.816-131.84-69.316-131.848 69.316  25.182-146.816-106.667-103.974 147.409-21.419z\\"></path></svg><ul></ul></div>"`;

exports[`Card should match the snapshot with a fact 1`] = `"<div class=\\"Card\\"><svg width=\\"2rem\\" enable-background=\\"new 0 0 426.667 426.667\\" viewBox=\\"0 0 426.667 426.667\\"><path d=\\"m213.333 10.441 65.916 133.576 147.418  21.419-106.667 103.974 25.173  146.816-131.84-69.316-131.848 69.316  25.182-146.816-106.667-103.974 147.409-21.419z\\" fill=\\"#f2f2f2\\"></path></svg><ul><li>name: test</li></ul></div>"`;

exports[`Card should match the snapshot with multiple facts 1`] = `"<div class=\\"Card\\"><svg width=\\"2rem\\" enable-background=\\"new 0 0 426.667 426.667\\" viewBox=\\"0 0 426.667 426.667\\"><path d=\\"m213.333 10.441 65.916 133.576 147.418  21.419-106.667 103.974 25.173  146.816-131.84-69.316-131.848 69.316  25.182-146.816-106.667-103.974 147.409-21.419z\\" fill=\\"#f2f2f2\\"></path></svg><ul><li>name: test1</li><li>origin: the matrix</li></ul></div>"`;
